cogs/games.py

```diff
@@ -0,0 +1,106 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright (C) idoneam (2016-2019)
+#
+# This file is part of Canary
+#
+# Canary is free software; you can redistribute it and/or modify it
+# under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# Canary is distributed in the hope that it will be useful, but
+# WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with Canary. If not, see <https://www.gnu.org/licenses/>.
+
+# discord-py requirements
+import discord
+from discord.ext import commands
+
+# Other utilities
+import random
+import re
+from .utils.dnd_roll import dnd_roll
+
+
+class Games(commands.Cog):
+    def __init__(self, bot):
+        self.bot = bot
+
+    @commands.command()
+    async def roll(self, ctx, arg: str = '', mpr: str = ''):
+        """
+        Performs a DnD-style diceroll
+        Supports modifiers, multiple rolls, any-sided dice
+        """
+        roll_pattern = re.compile(r'^(\d*)d(\d*)([+-]?\d*)$')
+        roll_cmd = roll_pattern.match(arg)
+        params = {
+            'sides': 20,    # Default parameters
+            'repeat': 1,
+            'mod': 0
+        }
+        if arg == 'safe':    # nice meme bro
+            await ctx.send('https://i.imgur.com/2icUGpc.png')
+            return
+        if arg == 'help' or arg == 'h':
+            helpmsg = discord.Embed(
+                title='DnD Dice Roller Help',
+                description='Usage: `[r]d[s][+m] [\'each\']`\n'
+                'Rolls an `s`-sided die `r` '
+                'times, with modifier `+-m`.\n'
+                'If `each` is specified, the '
+                'modifier is applied to each roll.\n'
+                'Rolls one 20-sided die by default',
+                colour=0x822AE0)
+            helpmsg.add_field(
+                name='Examples',
+                value='`d6` roll one 6-sided die\n'
+                '`3d` rolls 3 20-sided dice\n'
+                '`5d12-4` rolls 5 12-sided dice, subtracting 4 from the total\n'
+                '`5d+3 each` rolls 5 20-sided dice, adding 3 to each roll',
+                inline=False)
+            await ctx.send(embed=helpmsg)
+            return
+        elif roll_cmd != None:    # Applying some bounds on parameters
+            if roll_cmd.group(1) != '':
+                params['repeat'] = min(max(1, int(roll_cmd.group(1))), 10000)
+            if roll_cmd.group(2) != '':
+                params['sides'] = min(max(1, int(roll_cmd.group(2))), 100)
+            if roll_cmd.group(3) != '':
+                params['mod'] = min(max(-100, int(roll_cmd.group(3))), 100)
```

Comment:
Couple of things:

Could you use 

```python
if roll_cmd.group(1):
   # do thing
```
instead?

I see some duplicated code here:

- you're assigning a new value to a particular dictionary key

- that value is a call to `min(max(someint, int(roll_cmd.group(something))), someint)`

Can this be pulled out into a function? I find it would improve clarity about what the calls to `min()` and `max()` are for and would also make it more readable – I found it hard to disentangle the parens in these statements.

Reply:
if he refactors the clamping thing into a function, we might be able to refactor params as the following:
also, i'm not convinced params needs to be a dictionary - is it ever passed around as a unit?
```python
def clamp_or_default(val, min_val, max_val, default):
    """
    Enforces a minimum and maximum (closed) bound on an integer.
    Returns a default value if val is not an integer or false-y.
    """
    try:
        if val or isinstance(val, int):
            return min(max(min_val, int(val)), max_val)
    except ValueError:
        pass
    return default
repeat = clamp_or_default(roll_cmd.group(1), 1, 10000, default=20)
sides = clamp_or_default(roll_cmd.group(2), 1, 100, default=1)
mod = clamp_or_default(roll_cmd.group(3), -100, 100, default=0)
```

Reply:
we can put clamp_or_default in a util somewhere perhaps

Reply:
Resolved in ea625fe, 4198eb9, 646c4cb, 87ceaaf